posts/2021/07/moving-scipy-to-meson.md

```diff
@@ -0,0 +1,489 @@
+    ...
+    config = Configuration('_pocketfft', parent_package, top_path)
+    ext = config.add_extension('pypocketfft',
+                               sources=['pypocketfft.cxx'],
+                               depends=['pocketfft_hdronly.h'],
+                               include_dirs=include_dirs,
+                               language='c++')
+    ext._pre_build_hook = pre_build_hook
+```
+
+
+## Some key Meson design principles
+
+Meson is a well-designed build systems, and both the docs and a number of
+talks by Meson devs on YouTube do a great job at explaining that design. So I
+won't try and give a full picture here. However, there are a few things that
+are particularly important and would have helped me to fully grasp when I
+started on this project. So let's have a look at those.
+
+*Builds must be out-of-tree.* Meson will not allow building inside your existing
+source tree - and that includes code generation. There are good reasons for
+this, but when coming from `distutils` it may bite you. For example, if you
+have a script to generate Cython `.pyx` and `.pxd` files (quite common in
+SciPy), those files cannot be placed next to the template, they must go into
+a build directory (the one we chose with `meson setup <builddir>`).
+
+*Meson is not Turing-complete, and not extensible via APIs.* What this means is
+that if something is not supported in Meson itself, you cannot just write a
+bit of Python code to hack around the problem. Instead, you must add it in
+Meson itself. If that takes too long, just fork Meson and use your fork until
+your feature is merged upstream. This *seems* painful, but it guarantees that
+people don't just copy around changes from project to project and long-term
+maintainability deteriorates. Instead, the philosophy is to fix things once
+for all users.
+
+*All source files and targets must be list explicitly.* This means that if you
+build a single Python extension from 50 Fortran files, you must list all 50
+files names. This is not really a problem in practice, but it can be a little
+verbose. Using a few helper snippets to generate file listings in IPython
+saves time:
+```python
+>>> for f in sorted(os.listdir('.')):
+...     if f.endswith('.f'):
+...         print(f"  '{f}',")
+... 
+  'dgamln.f',
+  'zabs.f',
+  ...
+```
+
+*Meson has an imperative DSL with immutable objects.* The syntax of the DSL is
+inspired by Python, and is easy to read. Objects - which can be dependencies,
+generated files, built extensions, etc. - being immutable makes things easy
+to debug, but in some cases it can restrict what you can do. For example,
+this code for building all `scipy.sparse.csgraph` extensions is quite
+elegant, but the `foreach` pattern will not work if the `.pyx` files are
+generated. This is because they'd then be objects created by `custom_target`,
+and there is no syntax to give them unique names within the `foreach` loop -
+this cost me some time to figure out:
```

Comment:
Is there an open issue or a meson document that you could refer to with "for more information ..."

Reply:
Not that I know of. I also don't think it's a bug, it's just a consequence of how the DSL is designed.